test/common/webdriver_bidi.py

```diff
@@ -0,0 +1,453 @@
+
+
+JsonObject = dict[str, Any]
+
+
+class WebdriverError(RuntimeError):
+    pass
+
+
+@dataclass
+class LogMessage:
+    level: str  # like "info"
+    type: str  # usually "console"
+    timestamp: int
+    args: list[object]
+    text: str
+
+    def __init__(self, message_params: JsonObject) -> None:
+        self.level = message_params["level"]
+        self.type = message_params["type"]
+        self.timestamp = message_params["timestamp"]
+        self.args = message_params.get("args", [])
+        self.text = message_params["text"]
+
+    def __str__(self) -> str:
+        return f"> {self.level}: {self.text}"
+
+
+@dataclass
+class BidiSession:
+    ws_url: str
+    session_url: str
+    process: asyncio.subprocess.Process
+
+
+# Return port numbers that were free at the time of checking
+# They might be in use again by the time the function returns...
+def pick_ports(count: int) -> list[int]:
+    sockets: list[socket.socket] = []
+    ports: list[int] = []
+
+    for _ in range(count):
+        sock = socket.socket()
+        sock.bind(('127.0.0.1', 0))
+        sockets.append(sock)
+        ports.append(sock.getsockname()[1])
+
+    for s in sockets:
+        s.close()
+
+    return ports
+
+
+def jsquote(js: object) -> str:
+    return json.dumps(js)
+
+
+def unpack_value(raw: Any) -> Any:
+    """Convert a WebDriver type/value object into a plain Python object"""
+
+    if not isinstance(raw, dict):
+        return raw
+    try:
+        type_ = raw["type"]
+    except KeyError as e:
+        raise ValueError(f"No type in {raw!r}") from e
+    if type_ in ["undefined", "null"]:
+        return None
+    if type_ in ["string", "boolean", "number"]:
+        return raw["value"]
+    # serialized BiDi objects, don't touch
+    if type_ in ["function", "node", "window"]:
+        return raw
+    if type_ == "array":
+        return [unpack_value(v) for v in raw["value"]]
+    if type_ == "object":
+        obj = {}
+        for k, v in raw["value"]:
+            obj[k] = unpack_value(v["value"])
+        return obj
+    raise ValueError(f"Unknown type in {raw!r}")
+
+
+# FIXME: what do you want from me? https://docs.python.org/3/reference/datamodel.html#async-context-managers doesn't say
+class WebdriverBidi(contextlib.AbstractAsyncContextManager):  # type: ignore[type-arg]
+    http_session: aiohttp.ClientSession
+
+    def __init__(self, *, headless: bool = False) -> None:
+        self.headless = headless
+        self.last_id = 0
+        self.pending_commands: dict[int, asyncio.Future[JsonObject]] = {}
+        self.logs: list[LogMessage] = []
+        self.bidi_session: BidiSession | None = None
+        self.future_wait_page_load = None
+        self.top_context: str | None = None  # top-level browsingContext
+        self.context: str | None  # currently selected context (top or iframe)
+
+        self.homedir_temp = tempfile.TemporaryDirectory(prefix="cockpit-test-browser-home-")
```

Comment:
by the way: I'm still really very much not okay with this :)

Reply:
I'll retract my objection here: although I find this sort of "externally-visible effects due to garbage collection" stuff to be very much unpythonic, the authors of the tempfile module seem not to agree with me.  They've added code explicitly to handle this case cleanly, don't throw warnings about it, and even document it as being acceptable:

https://docs.python.org/3/library/tempfile.html

> It will be destroyed as soon as it is closed (including an implicit close when the object is garbage collected).

Nowhere is there a mention that you should explicitly call `.cleanup()`.  In fact, the docs state (emphasis mine):

> The directory **can** be explicitly cleaned up by calling the cleanup() method.

Reply:
I think the mental model here is something like "since the tempdir is only used inside of the program, it's not meant to be externally visible, so cleanup via GC is correct".  Of course, the one thing that derails this argument is that we *are* explicitly sharing this path with other processes (namely, the browser) but ... meh.  We're into the weeds at this point, so I'm not going to insist.  And I agree that there's a certain niceness to just creating it with the object and stopping to worry about it.

Reply:
I didn't touch this for now, please let me know if you still want me to change this (FTR: I've never called `.cleanup()` anywhere so far).

Reply:
I think it's fine.  It makes me cringe, but it's clearly the intended documented use of this API.  And it's much easier how it is now than what I was proposing.